test: use MiniWallet for mempool_package_onemore.py

theStack opened the pull request:
This PR enables one more of the non-wallet functional tests (mempool_package_onemore.py) to be run even with the Bitcoin Core wallet disabled by using the MiniWallet instead, as proposed in https://github.com/bitcoin/bitcoin/issues/20078. For this purpose helper methods `MiniWallet.{create,send}_self_transfer_multi` are introduced which serve as a replacement for `chain_transaction`. With this, it should be also quite straight-forward to change the larger related test `mempool_packages.py` to use MiniWallet.

theStack committed:
test: MiniWallet: add helper methods `{send,create}_self_transfer_multi`

theStack committed:
test: use MiniWallet for mempool_package_onemore.py

This test can now be run even with the Bitcoin Core wallet disabled.

MarcoFalke commented on test/functional/test_framework/wallet.py:
```diff
@@ -188,6 +188,38 @@ def send_to(self, *, from_node, scriptPubKey, amount, fee=1000):
 
+    def chain_tx(self, node, parent_txids, num_outputs=1, fee_per_output=1000):
+        """
+        Build and send a transaction that spends the given inputs and creates a
```
```suggestion
        Build and send a transaction that spends one output of the given txids and creates a
```

I think it only spends the first found (random) output

MarcoFalke replied:
Maybe it would be better to get rid of this and put the burden of coin selection on the caller? This would also allow the caller to decide whether the inputs are marked as spent or not if the call fails. (Other methods here also mark them as spent by default, even on failure)

theStack replied:
Yes, I agree. Reworked the helper to accept and return UTXOs rather than txids now (i.e. caller has to do coin selection), which makes the test even shorter.

maflcko reviewed (approved):
Concept ACK. Some questions/thoughts.

MarcoFalke commented on test/functional/test_framework/wallet.py:
```diff
@@ -188,6 +188,35 @@ def send_to(self, *, from_node, scriptPubKey, amount, fee=1000):
         txid = self.sendrawtransaction(from_node=from_node, tx_hex=tx.serialize().hex())
         return txid, 1
 
+    def chain_tx(self, node, utxos, num_outputs=1, fee_per_output=1000):
```
nit: (Some renames for symmetry)

```suggestion
    def send_self_transfer_multi(self, node_from, utxos, *, num_outputs=1, fee_per_output=1000):
```

maflcko reviewed (commented):
Left some more nits :sweat_smile:

theStack commented:
Thanks again for the valuable feedback @MarcoFalke! Force-pushed with the following suggested changes:
- removed the assertion for non-empty UTXO list (https://github.com/bitcoin/bitcoin/pull/24637#discussion_r832217228)
- renamed helper method and split it up into `create_self_transfer_multi`/`send_self_transfer_multi` for the symmetry with existing helpers (https://github.com/bitcoin/bitcoin/pull/24637#discussion_r832218529 and https://github.com/bitcoin/bitcoin/pull/24637#discussion_r832220737)
The PR description was also adapted accordingly.

MarcoFalke commented on test/functional/test_framework/wallet.py:
```diff
@@ -188,6 +188,41 @@ def send_to(self, *, from_node, scriptPubKey, amount, fee=1000):
         txid = self.sendrawtransaction(from_node=from_node, tx_hex=tx.serialize().hex())
         return txid, 1
 
+    def send_self_transfer_multi(self, from_node, utxos, **kwargs):
```
nit:

```suggestion
    def send_self_transfer_multi(self, *, from_node, utxos, **kwargs):
```

Also, "utxos" can be renamed to "inputs" or "utxos_to_spend".

MarcoFalke replied:
```suggestion
    def send_self_transfer_multi(self, **kwargs):
```

or maybe even just

MarcoFalke commented on test/functional/test_framework/wallet.py:
```diff
@@ -188,6 +188,41 @@ def send_to(self, *, from_node, scriptPubKey, amount, fee=1000):
+        """
+        tx = self.create_self_transfer_multi(from_node, utxos, **kwargs)
+        txid = self.sendrawtransaction(from_node=from_node, tx_hex=tx.serialize().hex())
+        return [self.get_utxo(txid=txid, vout=vout) for vout in range(len(tx.vout))]
```
nit: Maybe return a dict similar to `send_self_transfer`?

```suggestion
        return {"new_utxo":[self.get_utxo(txid=txid, vout=vout) for vout in range(len(tx.vout))], "tx": tx}
```

maflcko reviewed (commented):
(feel free to ignore)

theStack commented:
Force-pushed again with the following changes:
- all parameters for the newly introduced methods `{create,send}_self_transfer_multi` methods must be passed as keyword arguments now (to be consistent with `{create,self}_transfer` methods)
- `send_self_transfer_multi` returns a dictionary with several informations (txid, tx serialized as hex, tx as CTransaction instance, list of new UTXOs)
- introduced small helper `chain_tx` in the test in order to avoid repeatedly passing `from_node` arguments and being able to pass utxos and num_outputs as positional arguments; returns new UTXOs
- `getrawtransaction` RPC at the end of the test is not needed anymore, since we can directly access the `tx` field from the result dictionary

MarcoFalke commented on test/functional/test_framework/wallet.py:
```diff
@@ -188,6 +188,46 @@ def send_to(self, *, from_node, scriptPubKey, amount, fee=1000):
+        # adapt output amounts (use fixed fee per output)
+        inputs_value_total = sum([int(COIN * utxo['value']) for utxo in utxos_to_spend])
+        outputs_value_total = inputs_value_total - fee_per_output * num_outputs
+        for i in range(num_outputs):
+            tx.vout[i].nValue = outputs_value_total // num_outputs
```
nit: Haven't tried, but this might be possible to write shorter

```suggestion
        for o in tx.vout:
            o.nValue = outputs_value_total // num_outputs
```

MarcoFalke replied:
Done in #24623

maflcko reviewed (approved):
ACK 2b6dd4e75b3ad2daff553fde018fe4c8f1187878 💾

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 2b6dd4e75b3ad2daff553fde018fe4c8f1187878 💾
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh0jgwAlqAJ870kGVFw8j59DsDnOgazr4Q3gZs3ev7+QPUDwABrwzfJggjUWYy8
G+fBs+4b/buewrlpMpLHMoAza3n1ib4yiR4Ayebbj65vyVhRVERBncLMlbnYDU7c
VQR0Fkuz5FTza9P7BuIegO3JPOXxPnJi5AKg8xSsIY7vZfNvXuwn74zyZqI14hCm
iDG4JtKtyiYpL+Mct9UWD9mKBUlEW1KnESejKVj5eQTG5ww0TijlNCfQ7W+JTE3m
ndwqmbuPI5RM7YqytsKuTpPc584nqr/zM1/AuIGM7aEM8BdwTW5StJCL6AuPL1B+
wD3w7HVltleskkvmj2407xPZkiGanxzOE7L3Q5G0GiWDCJFYU6jqXbRoJ8ml/oPn
b8kILOTtLvEqRNry8bOGiOAxTGZG8VkdsvRtjxBE/HLGtgX/avlN9ELPPvLE6GKD
rY+isoIqlqqGGZfLegweVnCuLJVS0+1o8HJAgwPyhyejVXbcU1EACstphCNdMtd8
s9QdQ3hV
=mjSo
-----END PGP SIGNATURE-----
```


</details>
